### PR TITLE
[FIX] loading: do not commit when updating modules without testing

### PR DESCRIPTION
**Pull request by william-andre**
A commit is done between the `post-` and the `end-` steps. If there is an issue during the `end-`, some inconsistencies might happen. So we only commit when the tests are enabled to avoid users breaking the database by updating a module in the UI.

### CODE REVIEW

**Comment by william-andre** [comment]
@wtaferner FYI

**Comment by HydrionBurst** [comment]
I agree the current `commit` is added in a strange place
But we should `commit` the updated module immediately
For example when install moduleA and moduleB, after installing moduleA we commit.
So, if moduleB's installation fails, moduleA will should still be installed
Aslo for perforance, udpating a module may change too many data, so we commit the cursor to avoid making the transaction too heavy

But I am not sure what `post-` and `end-` you mean in your description
do you mean the `post-` migration and `end-` migration? because there are only `at_install` tests and `post_install` tests

Since it is for master branch, would you mind checking this PR of mine can solve your problem?
https://github.com/odoo/odoo/pull/121137

**Comment by wtaferner** [comment]
Yes, I do and no, post and pre is only triggered during init (install) not during a normal upgrade, right?

https://github.com/odoo/odoo/blob/a750e61544138a9d533f3391fbd2a179697fd7b4/addons/l10n_ch/migrations/11.1/end-migrate_update_taxes.py#L7

Here it is possible the database will raise a validation error introduced in later versions during/after upgrade or any other issue within the upgrade procedure at that stage.

This is an example code from Odoo core where it is implemented by now.

There might be a better hook where this should be executed, but as I was trying to follow the core developers of Odoo, I would have assumed that it is the current "supported" way to do so. I know it is not really by now, but for the sake of improving the situation we need to get rid or get the commit under full control otherwise it is just what commits normally do, destroying consistency if not used very wisely and being bullet-proof 😉

File: odoo/modules/loading.py [hunk 1]
```diff
@@ -272,6 +268,10 @@ def load_module_graph(env, graph, status=None, perform_checks=True,
         test_time = test_queries = 0
         test_results = None
         if tools.config.options['test_enable'] and (needs_update or not updating):
+            # need to commit any modification the module's installation or
+            # update made to the schema or data so the tests can run
+            # (separately in their own transaction)
```
Comment (rco-odoo) [inline odoo/modules/loading.py:273]:
```suggestion
12345678901234567890123456789012345678901234567890123456789012345678901234567890
            # commit any modification the module's installation or upgrade made
            # to the database so the tests can see them in their own transaction
```

**Review by rco-odoo — COMMENTED** [review]
The proposal kinda makes sense from a transactional point of view.
But for the sake of performance long transactions can become awfully slow.
@KangOl I would like to know your opinion on this.

File: odoo/modules/loading.py (hunk 1 above)
Comment (william-andre) [reply odoo/modules/loading.py:273]: > 12345678901234567890123456789012345678901234567890123456789012345678901234567890

confused

**Comment by wtaferner** [comment]
> I think usually the end- doesn't expect error, expecially in production

I agree, but think about it...the end stage is invoked by the module which is raising the version number in database without knowing what the outcome of the transaction will be, done at the end?

This is just plain wrong from a transactional standpoint...of course best to have no error and prepare perfectly but this is not how software works...so every commit is very dangerous if it is not 100% bullet proof.

A solution would be to rollback only the version number if there is a failure for a certain module in the end stage...then it could try again later based on that and would be safer for this specific use case at least.

Of course you can blame me or anyone who get an error in this stage, but I do not think it is how it should be.

**Disclaimer**: I have of course removed already what has caused the error, but made @william-andre and myself aware of this awkward issue. 

Curious what this will change or not in the end...thank you guys for discussing and trying to find a solution here.
Much appreciated 🙏🏼

**Comment by aj-fuentes** [comment]
I agree with @william-andre and @wtaferner that there is an issue here. This will become even more common now that the upgrade of taxes is implemented as local upgrade scripts in standard.

The change looks correct. I cannot say much about the performance penalties it would carry.  
Maybe we could implement some sort of warning with an extra go-ahead by the user when we detect the upgrade is going to trigger the upgrade of many modules (threshold TBD). Another idea would be to pre fetch a list of upgrade scripts to be run, if there is no end- one then go ahead an commit in-between modules. This would have the downside that _some_ upgrades would have the commits while others not, that's a big difference that could cause issues on its own.

@HydrionBurst the errors causes by unknown bugs in Odoo or wrong data in the DB are impossible to avoid. More so in upgrade scripts that tend to heavily manipulate data. Thus I agree with what was said here: there should be a way to avoid half-upgraded modules.

Note, resetting the module version back to what the upgrade started with is not a solution because most upgrade scripts won't work upon a half done upgrade.

To sum up: IMHO if this is the only way to fix it, I would prefer the performance hit than leaving the DB broken.

**Comment by william-andre** [comment]
A way to not hit the performances when `test_enabled` is set would be to still keep the previous commit, but guard it with `if tools.config.options['test_enable'] and (needs_update or not updating):`

**Comment by rco-odoo** [comment]
@wtaferner @aj-fuentes Please don't consider performance to be a minor issue in this case.
When a transaction becomes larger than a certain threshold, all subsequent SQL queries are _significantly_ slower.
And we are not talking about 10% slower, but _an order of magnitude_ slower.

It's not a coincidence that we split some upgrade operations on large tables into smaller transactions that are run in parallel (`parallel_execute()`).  The parallelization certainly helps, but the fact of making smaller transactions also helps.  And by the way, forget about transactional behavior in this case, as some parallel transactions may fail while other succeed.

I am not against this proposal, but I really don't have a clear intuition of what the result will be in terms of performance. I suggest to benchmark the branch on some realistic case, at least to get some idea.

**Comment by aj-fuentes** [comment]
> Please don't consider performance to be a minor issue in this case.

@rco-odoo Noted. I agree this should be benchmarked. 

I find confusing that this commit apparently was meant for running tests, and now we rely on it for better performance. TBH I feel I'm unable to fully approve this current patch beyond the "it looks correct".

**Comment by rco-odoo** [comment]
> I find confusing that this commit apparently was meant for running tests, and now we rely on it for better performance. TBH I feel I'm unable to fully approve this current patch beyond the "it looks correct".

I believe its only purpose was for running tests.  I don't think it was meant for performance.  But my guts tell me that we should be careful as it _might_ hurt performance.

I remember that in the past we removed quite a bunch of such wild commits.  I think there was a commit after every file being loaded.  If I am not mistaken, there are three commits left:
- one after each module installation/upgrade (this one, actually);
- two after uninstalling modules (one after deleting module data, and one after recreating dropped SQL views).

At the time we thought it was reasonable to have one transaction per module being installed/upgraded.  If I am not mistaken, the issue is that "end-upgrade" hooks are run at the very end, and are somewhat disconnected from their corresponding module upgrade.

**Comment by aj-fuentes** [comment]
I did a check and the only local end scripts we have in standard are all in accounting. Would it be possible to convert them to post scripts instead? (@clbr-odoo ?)

OTOH I don't see any in master. If they aren't needed anymore I think we can just keep them out of standard code and close this PR. @william-andre ?


> There might be a better hook where this should be executed, but as I was trying to follow the core developers of Odoo, I would have assumed that it is the current "supported" way to do so.

After carefully checking how it works. I think if we can do without end scripts for local upgrade that's always better. End scripts are mostly useful for major version upgrades when we can discard the whole DB if something fails. They are useful to perform some operation _after_ all modules are loaded. During major upgrades so many changes occur that some scripts can only, for example, remove a column at end because other modules need it for their upgrade. For upgrading a module in a production DB I would recommend using only pre/post scripts. Note that we can always put a post script in the last module that we need fully loaded/upgraded for the operation it performs.

**Comment by wtaferner** [comment]
Any news on this?
